### PR TITLE
Added new config system with support for contexts

### PR DESCRIPTION
**Pull request by severinson**
This PR introduces a new config system with support for contexts. This new system isn't complete and isn't used by any part of Armada yet. The intention is to merge this into master, perform changes to other parts of the code base required to support contexts, and then finishing implementing contexts. In addition, it cleans up the existing config system somewhat.

### CODE REVIEW

File: pkg/client/command_line.go
Comment (JamesMurkin) [inline pkg/client/command_line.go:8]: Why is this in its own group?

**Comment by severinson** [comment]
There is a failing e2e test related to the watch command. However, we need to wait until PR 792 is merged to work on it, since it may be fixed by PR 792 already, and any work on armadactl in this branch would introduce merge conflicts.
https://github.com/G-Research/armada/pull/792

File: pkg/client/command_line.go
Comment (d80tb7) [reply pkg/client/command_line.go:8]: If you're on goland then adding a file watcher that runs gofmt and go goimports on file save is awesome.

File: pkg/client/command_line.go
Comment (d80tb7) [inline pkg/client/command_line.go:19]: don't need the `at` here

File: pkg/client/command_line.go
Comment (d80tb7) [inline pkg/client/command_line.go:36]: I'd be tempted to log this.  It may save someone some time one day when the defaults file ahs been incorrectly deleted and armadactl doesn't work.

File: pkg/client/command_line.go
Comment (JamesMurkin) [inline pkg/client/command_line.go:19]: This is probably just a style thing, I find this hard to understand.

Here is another attempt, which may not be any better:
// LoadCommandlineArgsFromConfigFile loads armadactl config
// armadactl-defaults.yaml - From exePath, where exePath is the path to the armadactl executable
// armada config file           - From cfgFile or defaulting to $HOME/.armadactl
// These configs are then merged

File: pkg/client/command_line.go
Comment (JamesMurkin) [inline pkg/client/command_line.go:80]: Is this used?

File: pkg/client/command_line.go
Comment (d80tb7) [inline pkg/client/command_line.go:61]: again I'd log this

File: pkg/client/command_line.go
Comment (d80tb7) [inline pkg/client/command_line.go:134]: I'm not familiar with go logging, but in logging frameworks in other languages you wouldn't need to add all the `[armadaClientConfig.Validate]`s because the logging framework could add it from the logging context.  It may be that something is possible here.  If not I'd be tempted to make it a const string and add it via the format string, but I accept that's really a matter of taste!

**Comment by JamesMurkin** [comment]
Overall the above comments are minor. The main thing I'm interested in is, is this the structure we want?

```
	ApiVersion     string
	Users          map[string]*userConfig
	Clusters       map[string]*clusterConfig
	Contexts       map[string]*contextConfig
	CurrentContext string
```

How does this relate to the actual connection details structure?
https://github.com/G-Research/armada/blob/master/pkg/client/connection.go#L18


This does pretty much correlate with kubeconfig. I guess we are saying the contents of ApiConnectionDetails will be split between Users/Clusters?


Is `clusters` the right word here?

File: pkg/client/command_line.go
Comment (d80tb7) [inline pkg/client/command_line.go:192]: can we log which files we resolved or does viper do that for us?

**Review by d80tb7 — APPROVED** [review]

File: pkg/client/command_line.go
Comment (JamesMurkin) [reply pkg/client/command_line.go:36]: This does mean if you aren't setting an `armadactl-defaults.yaml` file. We'd log an error on every command.

Could we have a verbose logging option maybe?

Comment (d80tb7) [reply pkg/client/command_line.go:36]: so it doesn't need to be an error-debug would be fine.  I'd assumed that `armada-defaults` would be installed as part of the client install (as we're assuming it's next to the exe) and therefore should generally be there?

Comment (JamesMurkin) [reply pkg/client/command_line.go:36]: Its so package managers internal to companies can put a default there along with the installed exe/binary. 
 - So their users don't need to specify any armada config like api
 
 However if the user isn't using a package manager - or their company decides not to populate a armada-defaults, it is not expected to be present.

File: pkg/client/command_line.go
Comment (severinson) [reply pkg/client/command_line.go:8]: Just a mistake on my part; fixed

File: pkg/client/command_line.go
Comment (severinson) [reply pkg/client/command_line.go:36]: Later, I want to add a `view` command to `armadactl`, which prints the merged config. This command could perhaps be made to print the source files used to produce the merged config (as suggested below), which might be the right tradeoff between clarity and verbosity.

File: pkg/client/command_line.go
Comment (severinson) [reply pkg/client/command_line.go:80]: It will be used by the `armadactl view` command when added. For now I'm using it for debugging.

File: pkg/client/command_line.go
Comment (severinson) [reply pkg/client/command_line.go:61]: See response to previous comment on this.

File: pkg/client/command_line.go
Comment (severinson) [reply pkg/client/command_line.go:134]: :)

File: pkg/client/command_line.go
Comment (severinson) [reply pkg/client/command_line.go:200]: Doesn't seem like viper does it for us. Instead, I've added code that adds the found config files to a slice.

**Comment by severinson** [comment]
> Overall the above comments are minor. The main thing I'm interested in is, is this the structure we want?
> 
> ```
> 	ApiVersion     string
> 	Users          map[string]*userConfig
> 	Clusters       map[string]*clusterConfig
> 	Contexts       map[string]*contextConfig
> 	CurrentContext string
> ```
> 
> How does this relate to the actual connection details structure? https://github.com/G-Research/armada/blob/master/pkg/client/connection.go#L18
> 
> This does pretty much correlate with kubeconfig. I guess we are saying the contents of ApiConnectionDetails will be split between Users/Clusters?
> 
> Is `clusters` the right word here?

I made the struct like this because I felt it was the one most natural from the point of view of a user. As you say, `ApiConnectionDetails` needs to be split between users and clusters. I choose cluster since I think a user would find that more netural than, e.g., server. I think a user would say "let's submit this job to the Armada cluster", as opposed to "let's submit this job to the Armada server", for example. Further, each cluster has a server, which makes the config file structure logical.

Note that the structure is a bit different from kubeconfig. In kubeconfig, the users, for example, are stored as a list, where each entry (corresponding to a user) has a name field. However, I felt that storing users, clusters, contexts as maps from a name to the details of the entry was better since we have to require that names are unique between users/clusters/contexts, which we get automatically from a map but not from a list.

**Review by JamesMurkin — APPROVED** [review]

**Review by d80tb7 — APPROVED** [review]

**Pull request merged**
